barcode_utils.py

```diff
@@ -0,0 +1,166 @@
+            badge_num=badge_num,
+            barcode_num=barcode_num,
+        )
+        generated_lines.append(line)
+
+        # ensure that we haven't seen this value before
+        if barcode_num in seen_barcodes:
+            raise ValueError('COLLISION: generated a badge# that\'s already been seen')
+        else:
+            seen_barcodes.append(barcode_num)
+
+    return generated_lines
+
+
+def generate_barcode_from_badge_num(badge_num, event_id=None, salt=None, key=None):
+    event_id = config['secret']['barcode_event_id'] if not event_id else event_id
+    salt = config['secret']['barcode_salt'] if not salt else salt
+    key = bytes(config['secret']['barcode_key'],'ascii') if not key else key
+
+    # packed data going to be encrypted is:
+    # byte 1 - 8bit event ID, usually 1 char
+    # byte 2,3,4 - 24bit badge number
+
+    salted_val = badge_num + (0 if not salt else salt)
+
+    if salted_val > 0xFFFFFF:
+        raise ValueError("badge_number is too high " + str(badge_num))
+
+    data_to_encrypt = struct.pack('>BI', event_id, salted_val)
+
+    # remove the highest byte in that integer (2nd byte)
+    data_to_encrypt = bytearray([data_to_encrypt[0], data_to_encrypt[2], data_to_encrypt[3], data_to_encrypt[4]])
+
+    if len(data_to_encrypt) != 4:
+        raise ValueError("data to encrypt should be 4 bytes")
+
+    if len(key) != 10:
+        raise ValueError("key length should be exactly 10 bytes")
+
+    encrypted_string = _barcode_raw_encrypt(data_to_encrypt, key=key)
+
+    # check to make sure it worked.
+    decrypted = get_badge_num_from_barcode(encrypted_string, salt, key)
+    if decrypted['badge_num'] != badge_num or decrypted['event_id'] != event_id:
+        raise ValueError("didn't encode correctly")
+
+    # check to make sure this barcode number is valid for Code 128 barcode
+    verify_barcode_is_valid_code128(encrypted_string)
+
+    return encrypted_string
+
+
+def get_badge_num_from_barcode(barcode_num, salt=None, key=None):
+    salt = config['secret']['barcode_salt'] if not salt else salt
+    key = bytes(config['secret']['barcode_key'],'ascii') if not key else key
+
+    decrypted = _barcode_raw_decrypt(barcode_num, key=key)
+
+    result = dict()
+
+    result['event_id'] = struct.unpack('>B', bytearray([decrypted[0]]))[0]
+
+    badge_bytes = bytearray(bytes([0, decrypted[1], decrypted[2], decrypted[3]]))
+    result['badge_num'] = struct.unpack('>I', badge_bytes)[0] - salt
+
+    return result
+
+
+def verify_barcode_is_valid_code128(encrypted_string):
+    for c in encrypted_string:
+        if c not in code128._charset_b:
+            raise ValueError("contains a char not valid in a code128 barcode")
+
+
+def _barcode_raw_encrypt(value, key):
+    if len(value) != 4:
+        raise ValueError("invalid barcode input: needs to be exactly 4 bytes")
+
+    # skip32 generates 4 bytes output from 4 bytes input
+    _encrypt = True
+    skip32.skip32(key, value, _encrypt)
```

Comment:
> my only requirement for the encryption was simply NOT something based on XOR, just so I can avoid a bunch of discussions on how XOR isn't 'real' encryption, or whatever :)

This sentence sums up everything that Paul has ever said about barcodes, ever :P

I mean, we're literally saying, "Look I know this isn't meaningfully different from an XOR, but if I used an XOR then a bunch of the other geeks with strong feelings about barcodes will think we need something even more complicated than our already-pointless XOR, so I'm *writing even more complicated code* *_just to avoid an argument that would lead to code which is even more complicated than this_*."  Which as a tactic seems... okay, I guess.  But you should seriously link Paul to this exchange.